Queue and stack implementations

alevel-petrenko opened the pull request:
My solution for the queue and stack implementations.

Commit:
Initial commit with resources.

Commit:
added buttons and loading for films

Commit:
Small fixes in code

Commit:
changes in Queue practice

alevel-petrenko committed:
Changes in xaml code.

alevel-petrenko committed:
Updates in Queue window layout

alevel-petrenko committed:
Updates in VM

Commit:
Changed comments in the VM.

Commit:
Changed the way the allNumber collection initialize.

alevel-petrenko committed:
Made projects as dll.

Commit:
Added a lot of small fixes in comments.
Added image as static resource for icon (wip).

Commit:
Added Autofac.

Commit:
Added icon to the app.

Commit:
Added MvvmCross fw.

Commit:
Added Remove and Get command

Commit:
Added disabling buttons Show and Remove.

Commit:
Changed the app layout.

Commit:
Added stack collection.

Commit:
Changed logic of calling RaisePropertyChanged.
Added single ICustomCollection interface.
Changed file names.

Commit:
Added config creator for the collection.

Commit:
Added empty classes for tests.

Commit:
Added tests for the helper.

Commit:
Added tests for the queue collection.

Commit:
Added tests for the stack.

Commit:
Avalonia project was removed from the current branch.

AlEkSunder commented on QueueImplementation/Business.UnitTest/Queue/LinkedListQueueTest.cs:
```diff
@@ -0,0 +1,23 @@
+	/// </summary>
+	/// <owner>Anton Petrenko</owner>
+	[TestClass]
+	public sealed class LinkedListQueueTest : QueueTestBase
```
Why do the Array implementations contain more tests than Linked-List?

alevel-petrenko replied:
Because the logic of adding new element to the array is a bit complex than to the Linked list. Therefore, it was necessary to add more tests.

AlEkSunder replied:
But you have the same public interface and number of tests should be the same, isn't it?

alevel-petrenko replied:
The difference in tests is caused by code coverage result.
In order to have better coverage result I added these additional tests.

alevel-petrenko replied:
It's not possible to move this additional logic of tests to the some base Queue test class. because of differences in realization. 
So, I left tests in ArrayQueueTest and ArrayStackTest as they are.

AlEkSunder commented on QueueImplementation/Business/Helper/CollectionType.cs:
```diff
@@ -0,0 +1,33 @@
+    /// Represents the types of possible collections.
+    /// </summary>
+    /// <owner>Anton Petrenko</owner>
+    public enum CollectionType
```
As I remember, an array and a linked-lest types are basic structures of the Queues and Stack that are implemented.
So when you store all this stuff in the same enum you make it confusing.

alevel-petrenko replied:
Added separate enum for custom collections.

AlEkSunder commented on QueueImplementation/VM/CustomCollectionViewModel.cs:
```diff
@@ -0,0 +1,363 @@
+	/// Represents the view model used for managing the collection.
+	/// </summary>
+	/// <owner>Anton Petrenko</owner>
+	public sealed class CustomCollectionViewModel<T> : MvxViewModel
```
Why do we need the type if we don't use it?
Lets discuss.

alevel-petrenko replied:
Removed generic parameter.

AlEkSunder commented on QueueImplementation/VM/CustomCollectionViewModel.cs:
```diff
@@ -0,0 +1,363 @@
+		/// Holds the configuration creator.
+		/// </summary>
+		/// <owner>Anton Petrenko</owner>
+		private readonly CollectionConfigCreator<int> configCreator = new CollectionConfigCreator<int>();
```
Please move all default values initialization to the constructor.

alevel-petrenko replied:
Fixed

AlEkSunder commented on QueueImplementation/VM/CustomCollectionViewModel.cs:
```diff
@@ -0,0 +1,363 @@
+					return this.allNumbers;
+
+				if (this.Collection != null)
+				this.allNumbers = new ObservableCollection<int>(this.Collection);
```
Please align.

alevel-petrenko replied:
Fixed

AlEkSunder commented on QueueImplementation/VM/CustomCollectionViewModel.cs:
```diff
@@ -0,0 +1,363 @@
+		/// </summary>
+		/// <owner>Anton Petrenko</owner>
+		/// <value>The collection.</value>
+		public ICustomCollection<int> Collection
```
Don't see usages of this collection on the UI.
To be honest it is not clear for me the idea with two collections - Collection and AllNumbers.

alevel-petrenko replied:
Now the collection with all numbers is only one as it should be.

AlEkSunder commented on QueueImplementation/VM/CustomCollectionViewModel.cs:
```diff
@@ -0,0 +1,363 @@
+		/// </summary>
+		/// <owner>Anton Petrenko</owner>
+		/// <value>The collection of all numbers.</value>
+		public IEnumerable<int> AllNumbers
```
The property does not meet "generic" requirements. What if we want to store chars immediately?

alevel-petrenko replied:
Whole this view model works with the integer type.

AlEkSunder commented on QueueImplementation/VM/CustomCollectionViewModel.cs:
```diff
@@ -0,0 +1,363 @@
+		/// </summary>
+		/// <owner>Anton Petrenko</owner>
+		/// <value>The collection types.</value>
+		public string[] CollectionTypes => new string[] { "stack", "queue" };
```
Literals should not be used.

alevel-petrenko replied:
Enumeration is used for this purpose.

AlEkSunder commented on QueueImplementation/VM/CustomCollectionViewModel.cs:
```diff
@@ -0,0 +1,363 @@
+		/// </summary>
+		/// <owner>Anton Petrenko</owner>
+		/// <value>The type of the selected array linked list.</value>
+		public string SelectedArrayLinkedListType { get; set; }
```
Please add default values for Types and Structures.
Also please change name to more informative to make possible to separate why we need two these types. Like BasicDataStructureTypes and DataStructureTypes.

alevel-petrenko replied:
Names were changed to more understandable and according to the new enumeration.
But setting the default value to the appropriate collection types means that these types will use during configurating of dependencies. Therefore, user won't be able to change the main settings.

I'm about these dependencies
CollectionConfigCreator.InitializeCollection

AlEkSunder replied:
I meant that there are empty comboboxes on the UI and it is not good-looking.

alevel-petrenko replied:
I suppose that empty combobox/radio button are not big deal. In that way user would be 100% percent sure that he chose the necessary collection type.

alevel-petrenko replied:
The type None is default for the CustomCollectionType. The ui looks in this way now.
![image](https://user-images.githubusercontent.com/41588762/102223188-0b23d100-3eed-11eb-88cf-4b034f6660bc.png)

AlEkSunder commented on QueueImplementation/VM/CustomCollectionViewModel.cs:
```diff
@@ -0,0 +1,363 @@
+
+				return this.show;
+			}
+			set
```
I think setter is excessive for the commands. But to make controls inactive we should handle that using CanGet* delegate.

alevel-petrenko replied:
Fixed.

AlEkSunder replied:
But you also removed the caching of the commands. And this should be reverted.

alevel-petrenko replied:
Restored

AlEkSunder commented on QueueImplementation/Business/CustomCollectionConfig.cs:
```diff
@@ -0,0 +1,68 @@
+		/// </summary>
+		/// <owner>Anton Petrenko</owner>
+		/// <param name="selectedArrayLinkedListType">The selected type of collection (array/linked list).</param>
+		private static void ConfigureStack(string selectedArrayLinkedListType)
```
Could we pass here an enum-element?

alevel-petrenko replied:
Radio buttons and combo box are bound with enumeration now.

AlEkSunder commented on QueueImplementation/Business/CustomCollectionConfig.cs:
```diff
@@ -0,0 +1,68 @@
+		/// </summary>
+		/// <owner>Anton Petrenko</owner>
+		/// <param name="selectedArrayLinkedListType">The selected type of collection (array/linked list).</param>
+		private static void ConfigureQueue(string selectedArrayLinkedListType)
```
The same as below.

alevel-petrenko replied:
Fixed.

AlEkSunder commented on QueueImplementation/Business.UnitTest/Helper/CollectionConfigCreatorTest.cs:
```diff
@@ -0,0 +1,81 @@
+		/// <param name="selectedArrayLinkedListType">The type of the selected collection (array/linked list).</param>
+		/// <param name="expectedType">The type of expected collection.</param>
+		[TestMethod]
+		[DataRow("Queue", "Array", typeof(ArrayQueue<char>))]
```
Good point that you have got acquainted with attribute-approach.

AlEkSunder reviewed (changes requested):
Please add necessary changes and lets discuss this a little bit.

Commit:
First bunch of fixes.
Added separate enums and changed the logic of binding for combobox and radio buttons.

Commit:
Left only one collection to work with.

Commit:
Added ObservableCollection to properly update values on UI

Commit:
Restored the caching of add, remove and show commands.

Commit:
Added default choise 'None' for the CustomCollectionType.
